Questão 06, Lista 03 \

min z = -4x1 + 6x2 \
Sujeito a: \
5x1 + 2x2 ≥ 17 \
3x1 - 2x2 ≤ 11 \
x1 e x2 ≥ 0



Problema relaxado - Programação Linear (cvxopt.modeling)

In [1]:
from cvxopt.modeling import op, variable
from plotly import graph_objects as go

In [ ]:
# Definindo variáveis de decisão
x1 = variable(1, 'variavel x1')
x2 = variable(1, 'variavel x2')

# Definindo a FOB
fob = -4 * x1[0] + 6 * x2[0]

# Definindo as restrições
restricoes = []

restricoes.append(5 * x1[0] + 2 * x2[0] >= 17)
restricoes.append(3 * x1[0] - 2 * x2[0] <= 11)
restricoes.append(x1[0] >= 0)
restricoes.append(x2[0] >= 0)

# Montando o PL
problema = op(fob, restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
    print("Solução ótima encontrada.")
    print(f'O valor da fob é {-fob.value()[0]}')
    print(f'{x1.name} é de {x1[0].value()[0]}')
    print(f'{x2.name} é de {x2[0].value()[0]}')
elif problema.status == 'primal infeasible':
    print("O problema é inviável.")
elif problema.status == 'unbounded':
    print("O problema é ilimitado.")
else:
    print("Status desconhecido da otimização.")

optimal
Solução ótima encontrada.
O valor da fob é 14.666666666666668
variavel x1 é de 3.666666666666667
variavel x2 é de 0.0


Programação Linear inteira (cvxopt.glpk) \

min z = -4x1 + 6x2 \
Sujeito a: \
5x1 + 2x2 ≥ 17 \
3x1 - 2x2 ≤ 11 \
x1 ≥ 0 e inteiro \
x2 ≥ 0 e inteiro \

In [ ]:
from cvxopt.glpk import ilp
import numpy as np
from cvxopt import matrix

In [ ]:
c=matrix([-4,6],tc='d')
G=matrix([[5,2],[3,-2]],tc='d')
h=matrix([17,11],tc='d')

In [ ]:
(status, x)=ilp(c,G.T,h,I=set([0,1]))

In [ ]:
print(status)

LP relaxation is dual infeasible


Programação linear inteira (Pulp)

In [ ]:
!pip install pulp

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.3/14.3 MB 76.6 MB/s eta 0:00:00


In [ ]:
import pulp

In [ ]:
# Criação do problema de minimização
problem = pulp.LpProblem("Problema teste", pulp.LpMinimize)

/usr/local/lib/python3.10/dist-packages/pulp/pulp.py:1352: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


In [ ]:
# Variáveis de decisão
x1 = pulp.LpVariable("x1", lowBound=0, cat="Integer")
x2 = pulp.LpVariable("x2", lowBound=0, cat="Integer")

In [ ]:
# Função objetivo
problem += -4 * x1 + 6 * x2, "Minimização Total"

In [ ]:
# Restrições
problem += 5 * x1 + 2 * x2 >= 17, "restrição 1"
problem += 3 * x1 - 2 * x2 <= 11, "restrição 2"

In [ ]:
# Resolvendo o problema
problem.solve()

1

In [ ]:
# Exibindo os resultados
print(f"Status da solução: {pulp.LpStatus[problem.status]}")
print(f"O valor da fob é: {pulp.value(problem.objective)}")
print(f"O valor de x1 é: {x1.varValue}")
print(f"O valor de x2: {x2.varValue}")

Status da solução: Optimal
O valor da fob é: -10.0
O valor de x1 é: 4.0
O valor de x2: 1.0


Branch and Bound - Solução Relaxada: \
(x1,x2) = (3.666666666666667, 0) \
F = 14.666666666666668 \

---

Pegando x1 e criando 2 subproblemas: \

Subproblema A: \
min z = -4x1 + 6x2 \
Sujeito a: \
5x1 + 2x2 ≥ 17 \
3x1 - 2x2 ≤ 11 \
x1 <= 3 \
x1, x2 ≥ 0

Subproblema B: \
min z = -4x1 + 6x2 \
Sujeito a: \
5x1 + 2x2 ≥ 17 \
3x1 - 2x2 ≤ 11 \
x1 >= 4 \
x1, x2 ≥ 0


Subproblema A:

In [2]:
# Definindo variáveis de decisão
x1 = variable(1, 'variavel x1')
x2 = variable(1, 'variavel x2')

# Definindo a FOB
fob = -4 * x1[0] + 6 * x2[0]

# Definindo as restrições
restricoes = []

restricoes.append(5 * x1[0] + 2 * x2[0] >= 17)
restricoes.append(3 * x1[0] - 2 * x2[0] <= 11)
restricoes.append(x1[0] <= 3)
restricoes.append(x1[0] >= 0)
restricoes.append(x2[0] >= 0)

# Montando o PL
problema = op(fob, restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
    print("Solução ótima encontrada.")
    print(f'O valor da fob é {-fob.value()[0]}')
    print(f'{x1.name} é de {x1[0].value()[0]}')
    print(f'{x2.name} é de {x2[0].value()[0]}')
elif problema.status == 'primal infeasible':
    print("O problema é inviável.")
elif problema.status == 'unbounded':
    print("O problema é ilimitado.")
else:
    print("Status desconhecido da otimização.")

optimal
Solução ótima encontrada.
O valor da fob é 6.0
variavel x1 é de 3.0
variavel x2 é de 1.0


Subproblema B:

In [3]:
# Definindo variáveis de decisão
x1 = variable(1, 'variavel x1')
x2 = variable(1, 'variavel x2')

# Definindo a FOB
fob = -4 * x1[0] + 6 * x2[0]

# Definindo as restrições
restricoes = []

restricoes.append(5 * x1[0] + 2 * x2[0] >= 17)
restricoes.append(3 * x1[0] - 2 * x2[0] <= 11)
restricoes.append(x1[0] >= 4)
restricoes.append(x1[0] >= 0)
restricoes.append(x2[0] >= 0)

# Montando o PL
problema = op(fob, restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
    print("Solução ótima encontrada.")
    print(f'O valor da fob é {-fob.value()[0]}')
    print(f'{x1.name} é de {x1[0].value()[0]}')
    print(f'{x2.name} é de {x2[0].value()[0]}')
elif problema.status == 'primal infeasible':
    print("O problema é inviável.")
elif problema.status == 'unbounded':
    print("O problema é ilimitado.")
else:
    print("Status desconhecido da otimização.")

optimal
Solução ótima encontrada.
O valor da fob é 13.000000000000007
variavel x1 é de 4.0
variavel x2 é de 0.49999999999999895


No Subproblema A foi encontrada uma solução ótima com números inteiros:

(x1, x2) = (3, 1) \
FA = 6 \

No subproblema B foi encontrada uma solução, mas fracionada:

(x1, x2) = (4, 0.49999999999999895) \
FB = 13.000000000000007 \

Irei tentar encontrar outra solução ótima com um valor de fob mais alto, a partir do subproblema A, tendo em vista que a solução do subproblema A já é uma candidata a melhor solução ótima.

Subproblema B1

min z = -4x1 + 6x2 \
Sujeito a: \
5x1 + 2x2 ≥ 17 \
3x1 - 2x2 ≤ 11 \
x1 >= 4 \
x2 >= 1 \
x1, x2 ≥ 0

Subproblema B2

min z = -4x1 + 6x2 \
Sujeito a: \
5x1 + 2x2 ≥ 17 \
3x1 - 2x2 ≤ 11 \
x1 >= 4 \
x2 <= 0 \
x1, x2 ≥ 0


Subproblema B1

In [4]:
# Definindo variáveis de decisão
x1 = variable(1, 'variavel x1')
x2 = variable(1, 'variavel x2')

# Definindo a FOB
fob = -4 * x1[0] + 6 * x2[0]

# Definindo as restrições
restricoes = []

restricoes.append(5 * x1[0] + 2 * x2[0] >= 17)
restricoes.append(3 * x1[0] - 2 * x2[0] <= 11)
restricoes.append(x1[0] >= 4)
restricoes.append(x2[0] >= 1) # restricao subproblema B1
restricoes.append(x1[0] >= 0)
restricoes.append(x2[0] >= 0)

# Montando o PL
problema = op(fob, restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
    print("Solução ótima encontrada.")
    print(f'O valor da fob é {-fob.value()[0]}')
    print(f'{x1.name} é de {x1[0].value()[0]}')
    print(f'{x2.name} é de {x2[0].value()[0]}')
elif problema.status == 'primal infeasible':
    print("O problema é inviável.")
elif problema.status == 'unbounded':
    print("O problema é ilimitado.")
else:
    print("Status desconhecido da otimização.")

optimal
Solução ótima encontrada.
O valor da fob é 11.333333333333336
variavel x1 é de 4.333333333333334
variavel x2 é de 1.0


Subproblema B2

In [5]:
# Definindo variáveis de decisão
x1 = variable(1, 'variavel x1')
x2 = variable(1, 'variavel x2')

# Definindo a FOB
fob = -4 * x1[0] + 6 * x2[0]

# Definindo as restrições
restricoes = []

restricoes.append(5 * x1[0] + 2 * x2[0] >= 17)
restricoes.append(3 * x1[0] - 2 * x2[0] <= 11)
restricoes.append(x1[0] >= 4)
restricoes.append(x2[0] <= 0) # restricao subproblema B2
restricoes.append(x1[0] >= 0)
restricoes.append(x2[0] >= 0)

# Montando o PL
problema = op(fob, restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
    print("Solução ótima encontrada.")
    print(f'O valor da fob é {-fob.value()[0]}')
    print(f'{x1.name} é de {x1[0].value()[0]}')
    print(f'{x2.name} é de {x2[0].value()[0]}')
elif problema.status == 'primal infeasible':
    print("O problema é inviável.")
elif problema.status == 'unbounded':
    print("O problema é ilimitado.")
else:
    print("Status desconhecido da otimização.")

primal infeasible
O problema é inviável.


No subproblema B1 foi encontrada uma solução fracionada:

(x1, x2) = (4.333333333333334, 1) \
FB1 = 11.333333333333336 \

No subproblema B2 retornou que o problema é inviável.

Irei continuar a busca por outra solução ótima, a partir do subproblema B1.

Subproblema B1_1

min z = -4x1 + 6x2 \
Sujeito a: \
5x1 + 2x2 ≥ 17 \
3x1 - 2x2 ≤ 11 \
x1 >= 4 \
x2 >= 1 \
x1 >= 5 \
x1, x2 ≥ 0

Subproblema B1_2

min z = -4x1 + 6x2 \
Sujeito a: \
5x1 + 2x2 ≥ 17 \
3x1 - 2x2 ≤ 11 \
x1 >= 4 \
x2 >= 1 \
x1 <= 4 \
x1, x2 ≥ 0

Subproblema B1_1:

In [6]:
# Definindo variáveis de decisão
x1 = variable(1, 'variavel x1')
x2 = variable(1, 'variavel x2')

# Definindo a FOB
fob = -4 * x1[0] + 6 * x2[0]

# Definindo as restrições
restricoes = []

restricoes.append(5 * x1[0] + 2 * x2[0] >= 17)
restricoes.append(3 * x1[0] - 2 * x2[0] <= 11)
restricoes.append(x1[0] >= 4)
restricoes.append(x2[0] >= 1) # restricao subproblema B1
restricoes.append(x1[0] >= 5) # restricao subproblema B1_1
restricoes.append(x1[0] >= 0)
restricoes.append(x2[0] >= 0)

# Montando o PL
problema = op(fob, restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
    print("Solução ótima encontrada.")
    print(f'O valor da fob é {-fob.value()[0]}')
    print(f'{x1.name} é de {x1[0].value()[0]}')
    print(f'{x2.name} é de {x2[0].value()[0]}')
elif problema.status == 'primal infeasible':
    print("O problema é inviável.")
elif problema.status == 'unbounded':
    print("O problema é ilimitado.")
else:
    print("Status desconhecido da otimização.")

optimal
Solução ótima encontrada.
O valor da fob é 8.000000000000005
variavel x1 é de 4.999999999999999
variavel x2 é de 1.9999999999999984


Subproblema B1_2

In [7]:
# Definindo variáveis de decisão
x1 = variable(1, 'variavel x1')
x2 = variable(1, 'variavel x2')

# Definindo a FOB
fob = -4 * x1[0] + 6 * x2[0]

# Definindo as restrições
restricoes = []

restricoes.append(5 * x1[0] + 2 * x2[0] >= 17)
restricoes.append(3 * x1[0] - 2 * x2[0] <= 11)
restricoes.append(x1[0] >= 4)
restricoes.append(x2[0] >= 1) # restricao subproblema B1
restricoes.append(x1[0] >= 4) # restricao subproblema B1_2
restricoes.append(x1[0] >= 0)
restricoes.append(x2[0] >= 0)

# Montando o PL
problema = op(fob, restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
    print("Solução ótima encontrada.")
    print(f'O valor da fob é {-fob.value()[0]}')
    print(f'{x1.name} é de {x1[0].value()[0]}')
    print(f'{x2.name} é de {x2[0].value()[0]}')
elif problema.status == 'primal infeasible':
    print("O problema é inviável.")
elif problema.status == 'unbounded':
    print("O problema é ilimitado.")
else:
    print("Status desconhecido da otimização.")

optimal
Solução ótima encontrada.
O valor da fob é 11.333333333333336
variavel x1 é de 4.333333333333334
variavel x2 é de 1.0


No subproblema B1_1 foi encontrada uma solução fracionada:

(x1, x2) = (4.999999999999999, 1.9999999999999984) \
FB1_1 = 8.000000000000005 \

No subproblema B1_2 foi encontrada uma solução fracionada:

(x1, x2) = (4.333333333333334, 1) \
FB1_2 = 11.333333333333336 \

Por enquanto, finalizarei a busca, enfatizando que no momento a melhor solução é subproblema A.

